x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilege.java

```diff
@@ -5,189 +5,16 @@
-            entry("create_snapshot", CREATE_SNAPSHOT),
-            entry("manage_ilm", MANAGE_ILM),
-            entry("read_ilm", READ_ILM),
-            entry("manage_slm", MANAGE_SLM),
-            entry("read_slm", READ_SLM));
-
-    private static final ConcurrentHashMap<Set<String>, ClusterPrivilege> CACHE = new ConcurrentHashMap<>();
-
-    private ClusterPrivilege(String name, String... patterns) {
-        super(name, patterns);
-    }
-
-    private ClusterPrivilege(String name, Automaton automaton) {
-        super(Collections.singleton(name), automaton);
-    }
-
-    private ClusterPrivilege(Set<String> name, Automaton automaton) {
-        super(name, automaton);
-    }
-
-    public static ClusterPrivilege get(Set<String> name) {
-        if (name == null || name.isEmpty()) {
-            return NONE;
-        }
-        return CACHE.computeIfAbsent(name, ClusterPrivilege::resolve);
-    }
-
-    private static ClusterPrivilege resolve(Set<String> name) {
-        final int size = name.size();
-        if (size == 0) {
-            throw new IllegalArgumentException("empty set should not be used");
-        }
-
-        Set<String> actions = new HashSet<>();
-        Set<Automaton> automata = new HashSet<>();
-        for (String part : name) {
-            part = part.toLowerCase(Locale.ROOT);
-            if (ACTION_MATCHER.test(part)) {
-                actions.add(actionToPattern(part));
-            } else {
-                ClusterPrivilege privilege = VALUES.get(part);
-                if (privilege != null && size == 1) {
-                    return privilege;
-                } else if (privilege != null) {
-                    automata.add(privilege.automaton);
-                } else {
-                    throw new IllegalArgumentException("unknown cluster privilege [" + name + "]. a privilege must be either " +
-                            "one of the predefined fixed cluster privileges [" +
-                            Strings.collectionToCommaDelimitedString(VALUES.entrySet()) + "] or a pattern over one of the available " +
-                            "cluster actions");
-                }
-            }
-        }
-
-        if (actions.isEmpty() == false) {
-            automata.add(patterns(actions));
-        }
-        return new ClusterPrivilege(name, Automatons.unionAndMinimize(automata));
-    }
-
-    public static Set<String> names() {
-        return Collections.unmodifiableSet(VALUES.keySet());
-    }
+/**
+ * This interface represents a privilege that is used to control access to cluster level actions.
+ */
+public interface ClusterPrivilege {
+    /**
+     * Uses {@link ClusterPermission.Builder} to add predicate that later can be used to build a {@link ClusterPermission}.
+     * @param builder {@link ClusterPermission.Builder}
+     * @return an instance of {@link ClusterPermission.Builder}
+     */
+    ClusterPermission.Builder buildPermission(ClusterPermission.Builder builder);
```

Comment:
We had a bunch of discussions about "implies", "privileges" and "permissions". Given this interface, it is impossible for a `PermissionCheck` to say if a `ClusterPrivilege` is implied or not.
I suggest we enrich this interface and have an `impliedBy(ClusterPermission clusterPermission)` or `impliedBy(PermissionCheck check)` method.

This is a suggestion for our future selves, not to be treated in this PR, ofc!